File: app/src/api/calendarService.ts
import { invoke } from '@tauri-apps/api/core';
import type { CalendarEvent, CalendarEventDraft } from '../types/calendar';

/**
 * Слой данных календаря поверх Tauri.
 *
 * React не знает о сервере, транспорте или токене — это ответственность
 * Rust-слоя (см. app/src-tauri/src/calendar.rs), который сам подставляет
 * токен сессии в запросы к backend'у.
 */

const ACCENT_PINK = '#ffb6d2';
const ACCENT_VIOLET = '#b478ff';
const STATUS_GREEN = '#3ecf6e';
const STATUS_AMBER = '#ffbd2e';
const STATUS_BLUE = '#4fc3f7';
const STATUS_RED = '#ff5252';
const STATUS_TEAL = '#26a69a';
const STATUS_ORANGE = '#ff7043';
const NEUTRAL_GRAY = '#9e9e9e';
const STATUS_INDIGO = '#5c6bc0';
const STATUS_LIME = '#d4e157';
const STATUS_CYAN = '#26c6da';
const STATUS_BROWN = '#8d6e63';

/** Палитра цветов событий, доступная в UI. Первые два связаны с токенами бренда. */
export const EVENT_COLORS = [
  ACCENT_PINK, 
  ACCENT_VIOLET, 
  STATUS_GREEN, 
  STATUS_AMBER, 
  STATUS_BLUE,
  STATUS_RED,
  STATUS_TEAL,
  STATUS_ORANGE,
  NEUTRAL_GRAY,
  STATUS_INDIGO,
  STATUS_LIME,
  STATUS_CYAN,
  STATUS_BROWN
] as const;

export async function getEvents(): Promise<CalendarEvent[]> {
  return await invoke<CalendarEvent[]>('get_calendar_events');
}

export async function createEvent(draft: CalendarEventDraft): Promise<CalendarEvent> {
  return await invoke<CalendarEvent>('create_calendar_event', { draft });
}

export async function updateEvent(event: CalendarEvent): Promise<CalendarEvent> {
  return await invoke<CalendarEvent>('update_calendar_event', { event });
}

export async function deleteEvent(id: string): Promise<void> {
  await invoke<void>('delete_calendar_event', { id });
}

File: mobile/ios/Loza/Services/CalendarService.swift
//
//  CalendarService.swift
//  Loza
//
//  Mirrors app/src/api/calendarService.ts: the event color palette and
//  the CRUD calls, now hitting the real backend directly (through
//  LozaAPIClient) instead of going through a Tauri command layer.
//
//  On iOS, calendar mutations prefer the live WebSocket (/ws/app) so the
//  server can broadcast create/update/delete pushes to sibling connections
//  (e.g. a desktop client of the same user). If the socket isn't connected
//  (backgrounded, not yet authenticated, etc.), we transparently fall back to
//  a plain HTTP round-trip — identical to how the Tauri client's ws_client.rs
//  falls back to direct HTTP when the socket is down.

import SwiftUI

enum CalendarPalette {
    /// Same 13 colors, same order, as EVENT_COLORS in calendarService.ts.
    static let colors: [Color] = [
        Color(hexString: "#ffb6d2"), // accent pink
        Color(hexString: "#b478ff"), // accent violet
        Color(hexString: "#3ecf6e"), // status green
        Color(hexString: "#ffbd2e"), // status amber
        Color(hexString: "#4fc3f7"), // status blue
        Color(hexString: "#ff5252"), // status red
        Color(hexString: "#26a69a"), // status teal
        Color(hexString: "#ff7043"), // status orange
        Color(hexString: "#9e9e9e"), // neutral gray
        Color(hexString: "#5c6bc0"), // status indigo
        Color(hexString: "#d4e157"), // status lime
        Color(hexString: "#26c6da"), // status cyan
        Color(hexString: "#8d6e63"), // status brown
    ]
}

enum CalendarService {
    private static func credentials() async throws -> (URL, String) {
        guard let baseURL = await ServerConfig.shared.baseURL else {
            throw AuthError.noServerConfigured
        }
        guard let token = await SessionStore.shared.session?.token else {
            throw AuthError.invalidCredentials
        }
        return (baseURL, token)
    }

    // ─── Event read ────────────────────────────────────────────────────────────

    static func getEvents() async throws -> [CalendarEvent] {
        // WS-first: calendar.get over /ws/app. Falls back to HTTP GET for a
        // quick initial load or when the socket is still connecting.
        if StatusSocket.shared.isConnected {
            do {
                let dtos = try await StatusSocket.shared.wsGetCalendarEvents()
                return dtos.map(CalendarMapper.map)
            } catch {
                // Fall through to HTTP if the WS request failed for any reason
                // (timeout, parse error, server reject). HTTP is the source of
                // truth and always returns the current state.
            }
        }
        let (baseURL, token) = try await credentials()
        let dtos = try await LozaAPIClient.shared.getCalendarEvents(baseURL: baseURL, token: token)
        return dtos.map(CalendarMapper.map)
    }

    // ─── Event write ───────────────────────────────────────────────────────────

    static func createEvent(_ draft: CalendarEventDraft) async throws -> CalendarEvent {
        let dto = CalendarMapper.draftDTO(from: draft)
        if StatusSocket.shared.isConnected {
            do {
                let created = try await StatusSocket.shared.wsCreateCalendarEvent(draft: dto)
                return CalendarMapper.map(created)
            } catch {
                // Fall back to HTTP.
            }
        }
        let (baseURL, token) = try await credentials()
        let created = try await LozaAPIClient.shared.createCalendarEvent(baseURL: baseURL, token: token, draft: dto)
        return CalendarMapper.map(created)
    }

    static func updateEvent(_ event: CalendarEvent) async throws -> CalendarEvent {
        let dto = CalendarMapper.dto(from: event)
        if StatusSocket.shared.isConnected {
            do {
                let updated = try await StatusSocket.shared.wsUpdateCalendarEvent(event: dto)
                return CalendarMapper.map(updated)
            } catch {
                // Fall back to HTTP.
            }
        }
        let (baseURL, token) = try await credentials()
        let updated = try await LozaAPIClient.shared.updateCalendarEvent(baseURL: baseURL, token: token, event: dto)
        return CalendarMapper.map(updated)
    }

    static func deleteEvent(_ id: String) async throws {
        if StatusSocket.shared.isConnected {
            do {
                try await StatusSocket.shared.wsDeleteCalendarEvent(id: id)
                return
            } catch {
                // Fall back to HTTP.
            }
        }
        let (baseURL, token) = try await credentials()
        try await LozaAPIClient.shared.deleteCalendarEvent(baseURL: baseURL, token: token, id: id)
    }
}
